src/codehelp/codehelp.vala

```diff
@@ -20,6 +20,272 @@
  * Code help utilities that don't belong to any specific class
  */
 namespace Vls.CodeHelp {
+
+    errordomain FormattingError {
+        CONFIGURATION, SPAWN_ERROR, FORMATTING_ERROR
+    }
+
+    Lsp.TextEdit format (Lsp.FormattingOptions options, Vala.SourceFile source) throws FormattingError {
+        File config;
+        try {
+            config = CodeHelp.get_uncrustify_config (options);
+        } catch (Error e) {
+            throw new FormattingError.CONFIGURATION ("Configuration: " + e.message);
+        }
+        // SEARCH_PATH_FROM_ENVP does not seem to be available even in quite fast distros like Fedora 35
+        var launcher = new SubprocessLauncher (SubprocessFlags.STDERR_PIPE | SubprocessFlags.STDOUT_PIPE | SubprocessFlags.STDIN_PIPE);
+        var env_vars = Environ.get ();
+        for (var i = 0; i < env_vars.length; i++) {
+            var env = env_vars[i];
+            // Patch the PATH variable to include some other "standard paths"
+            if (env.has_prefix ("PATH=") && env.contains ("/usr/bin")) {
+                env_vars[i] = env + ":/usr/local/bin";
+            }
+        }
+        launcher.set_environ (env_vars);
+        Subprocess subprocess;
+        try {
+            subprocess = launcher.spawn ("uncrustify", "-c", config.get_path (), "-l", "vala", null);
+            uint8[] contents;
+            FileUtils.get_data (source.filename, out contents);
```

Comment:
I think you can use `source.content` instead of reading the file. This means that we're formatting what the user has just typed, not necessarily what the user has saved.